### ISSUE TITLE
[方針] C++高速Napoleon simulation / rollout engine

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#124]
## 親方針
- #14 playing RL
- #121 dynamic batched CUDA rollout

## 背景
現行playing rolloutでは、ゲームルール・状態遷移・observation生成・legal action生成をTypeScript / Node.jsで実行し、playing policyのNN forwardのみONNX Runtime CUDAへ渡している。

#121でdynamic batchingを導入した結果、同一2000 games / mixed rosterでONNX `session.run()` 回数は大幅に削減できた。

実機RTX4060では:

```text
baseline CUDA batch=1 / workers=4
  rollout: 11.20 sec
  session.run: ~80,000

concurrency=32 / maxBatch=32
  rollout: 18.45 sec
  session.run: 9,316
  mean batch: 8.59

concurrency=64 / maxBatch=64
  rollout: 18.67 sec
  session.run: 4,761
  mean batch: 16.80

concurrency=128 / maxBatch=128
  rollout: 19.51 sec
  session.run: 2,275
  mean batch: 35.16

concurrency=256 / maxBatch=256
  rollout: 21.57 sec
  session.run: 1,143
  mean batch: 69.99
```

GPU呼び出し回数は約70分の1まで減ったが、rollout全体は高速化しなかった。

これはCUDA pathでshared in-process ONNX sessionを使う一方、game simulation自体は1つのNode.js process / event loop上で進むため、CPU側simulationが主要コストとして残っている可能性を示している。

## 現状の強化学習結果

高速化方針の判断材料として、現行Actor-Criticで長期runを実施済み。

### 条件

```text
algorithm: Actor-Critic v1
iterations: 750
games / iteration: 2,000
total training games: 1,500,000
training batch size: 2,048
learning rate: 1e-5
value loss coefficient: 0.5
epochs: 1

training roster:
  current policy x3
  RuleBased x1
  frozen rl-v740 x1

evaluation:
  candidate vs RuleBased x4
  100 seeds x 5 seat rotations = 500 games
  every 10 generations
```

### 評価結果

代表点:

```text
generation   winRate
v0           57.0%
v10          57.2%
v100         56.0%
v300         55.6%
v500         55.6%
v650         54.2%
v750         55.0%
```

v10の57.2%はv0との差が500 games中1勝分のみであり、明確な改善とは判断しない。

750 generationsを通して長期的な改善は確認できず、むしろ後半は55%前後へ緩やかに低下した。

一方で:

```text
contract success rate: おおむね 0.33 前後
average point cards:   おおむね 9.5 前後
```

で大きな崩壊は見られない。

### 現時点の解釈

この結果だけでActor-Criticそのものを否定しない。

今後切り分ける候補には少なくとも以下がある。

- training roster
- opponent distribution
- self-play比率
- frozen policy比率
- reward設計
- learning rate
- batch size
- advantage / critic設計
- algorithm（将来的なPPO / GAE等）

特に現行runではtraining rosterの3/5がcurrent policyである一方、評価はcandidate vs RuleBased x4であり、training environmentとevaluation environmentに差がある。

次の有力な実験候補として:

```text
current x1
remaining 4 seats:
  RuleBased または frozen v740 をweighted random sampling
current seatは均等rotation
```

を想定している。

ただし、これらの条件探索を現行速度のまま多数回すとexperiment costが高い。

したがって **次に学習条件を細かく最適化する前に、AI simulation / rollout基盤そのものを高速化し、iteration単価を下げる** ことを優先する。

## 目的
ナポレオンAIの学習・評価・検証を高速化するための **C++ native simulation / rollout runtime** を導入する。

これはブラウザゲーム本体や正式なゲーム実行基盤をC++へ移行するものではない。

正式なルール実装・Webゲーム側のsource of truthは引き続きTypeScript `packages/game-core` とする。

C++側は **AI学習・AI評価・AI開発専用の高速simulation基盤** と位置づける。

このruntime上では、学習開発に必要な複数用途を扱える構造とする。

- RL training rollout
- policy evaluation
- benchmark
- tournament / policy-vs-policy比較
- opponent pool / mixed roster
- self-play
- randomized regression / stress test
- TypeScript game-coreとのdifferential test
- 将来的なsearch / MCTS等、AI開発上の高速simulation用途

## 用途境界
このC++ runtimeは **AI学習・AI評価・AI検証のためだけに使用する**。

以下には使用しない。

- ブラウザ上の実ゲーム進行
- Web UIから直接利用するゲームエンジン
- Fastify等の通常ゲームサーバーにおける正式game state管理
- `packages/game-core` の置き換え
- 人間プレイヤー向けゲーム進行のsource of truth
- 本番Web runtimeのルール判定

C++側とTypeScript側で差異が生じた場合、正式ルールのsource of truthはTypeScript `packages/game-core` とし、C++側を修正する。

C++側は速度優先の二重実装であるため、TypeScriptとのdifferential/parity testを必須とする。

## 基本設計
3層に分ける。

### Layer 1: Napoleon Core
AI / dataset / RL非依存の高速simulation用ゲーム状態・ルール。

最低限:
- `GameState`
- action / card / player representation
- legal action generation
- `step()` / state transition
- trick resolution
- joker / same2等playing phaseルール
- terminal detection
- winner / result
- deterministic RNG / seed semantics

C++ hot pathでは可能な範囲で:
- fixed-size struct / array
- enum / integer index
- bitset / bitmask
- allocation削減
- string card id依存の排除
- object graph / GC不要

を優先する。

ただし独自ルール簡略化は不可。正式ルールとの一致を優先する。

### Layer 2: Simulation Runtime
多数gameを同時管理する高速scheduler。

概念的には各gameを:

```text
RUNNABLE
WAITING_AGENT_A
WAITING_AGENT_B
...
FINISHED
```

のように管理し、RuleBased等CPUだけで進められる区間はC++内で進め、NN inferenceが必要なgameだけpolicy単位でbatch request化する。

目標イメージ:

```text
C++ Simulation Runtime
  ├ game 1
  ├ game 2
  ├ ...
  └ game N
        ↓
 policyごとの inference request queue
        ↓
 batched CUDA inference
        ↓
 action batch
        ↓
 C++側でsimulation再開
```

Node ↔ C++ を1 decisionごとに往復する設計は避ける。

最低限256 logical games程度を扱え、将来的に512 / 1024等も試せるよう固定上限へ強く依存しない。

logical game concurrencyとGPU batch sizeは別設定とする。

### Layer 3: AI Development Applications
Core / Runtime上にAI学習・評価用途のapplicationを載せる。

#### RL rollout
現行tensor-ready self-play dataset contractと互換なtraining data生成。

#### evaluation
candidate vs RuleBased / frozen policy / opponent pool等を高速評価。

#### tournament
複数policy / RuleBasedを大量対戦させ比較可能にする。

#### benchmark
simulation / inference / serialization / throughputを測定。

#### differential test
TypeScript game-coreをoracleとしてC++ correctnessを検証。

## Agent interface
C++ simulation coreからagent実装を分離する。

最低限以下を扱える抽象化を用意する。

```text
RuleBased
ONNX policy
Frozen ONNX policy
Current policy
```

runtime側はseatへagentを割り当てる。

例:

```text
seat0: current
seat1: rule
seat2: v740
seat3: rule
seat4: v740
```

self-playは:

```text
current current current current current
```

評価は:

```text
candidate rule rule rule rule
```

のようなroster設定として表現できるようにする。

## Roster / opponent pool
特定の学習用途をハードコードしない。

最低限:
- fixed roster
- current seat rotation
- weighted opponent pool
- seatごとのindependent sampling
- self-play

を表現可能な設計にする。

次の実用ユースケースとしては:

```text
current x1
残り4席を各seat独立に
  RuleBased 50%
  frozen v740 50%
current seatは均等rotation
```

を想定する。

ただし50/50やv740をcoreへハードコードしない。

## inference backend
初期実装では以下いずれでもよい。

1. C++ ONNX Runtime CUDAを直接利用
2. C++ simulation runtime + shared inference broker

ただしhot pathでNodeとのper-decision IPCを大量発生させない。

複数ONNX policyを同一rolloutで使用可能にし、policyごとに独立したbatch queue / sessionとして扱えること。

例:

```text
current requests -> current CUDA batch
v740 requests    -> v740 CUDA batch
RuleBased        -> C++ CPU
```

## observation / model input
Game Coreそのものを特定model schemaへ密結合させない。

構造は:

```text
Game Core
  ↓
public state / observation
  ↓
AI application / encoder
  ├ playing model input v2 (6246)
  ├ evaluation stats
  └ replay / debug trace
```

とする。

現行playing model input schema v2（6246 features）との完全parityはRL rollout application側で必須。

## dataset出力
RL rollout applicationでは現行playing self-play tensor-ready binary contractと互換なtraining dataを生成可能にする。

最低限:
- modelInput
- legalPlayMask
- selectedCardIndex
- behaviorLogProbability
- terminalReward
- role / seed / step等

#120のephemeral raw cache lifecycleと接続可能にする。

Dataset serializationはGame Coreへ直接埋め込まずapplication側へ分離する。

## TypeScript game-coreとのparity
C++は高速simulation用の二重実装になるため、differential/parity testを必須とする。

同一seed / 同一action sequenceについて最低限:
- legal actions
- current player
- hand contents
- trick state
- completed trick winner
- game termination
- winner
- playing observation
- model input v2

がTypeScript `packages/game-core` / existing rolloutと一致すること。

randomized differential testを多数seedで実行可能にする。

C++側の独自簡略ルールは不可。

## observability / benchmark
最低限:
- total elapsed
- simulation CPU elapsed
- inference elapsed
- serialization/write elapsed
- games/sec
- decisions/sec
- inference request count
- session.run count
- mean/max batch size
- policy別batch statistics

を計測可能にする。

RTX4060上で現行TypeScript runtimeと比較する。

baseline:

```text
CUDA batch=1 / TS workers=4: 11.20 sec / 2000 games
current batched TS path: 18-22 sec / 2000 games
```

最終的にはend-to-end 1 iterationの時間も比較する。

## 実装方針
このIssue #124自体は実装Issueではなく、C++高速simulation基盤全体の親方針Issueとする。

実装はこのIssueをsource of truthとする複数の `[実装]` sub-Issueへ分割して進める。

想定する依存順:

1. C++ Napoleon Core骨格 + TypeScript differential harness
2. playing phase state transition / legal action / trick resolution parity
3. observation / playing model input v2 parity
4. generic Agent / Roster interface
5. deterministic multi-game Simulation Runtime
6. RuleBased agent接続
7. ONNX agent + policy別batched CUDA inference
8. RL tensor dataset application
9. evaluation / benchmark application
10. orchestrator integration

## 最初の実用ユースケース
C++基盤の最初のproduction useとして:

```text
current x1
remaining seats:
  RuleBased or frozen v740 weighted random pool
current seat rotation
```

による高速RL rolloutを通す。

ただしengine自体はこのroster専用にしない。

## 今回やらない
- ブラウザゲーム本体へのC++導入
- Web UIからC++ runtimeを直接利用すること
- Fastify等の通常ゲームサーバーへのC++ game state導入
- TypeScript `packages/game-core`削除・置換
- C++を正式ルールのsource of truthにすること
- 人間対戦・人間プレイ用runtimeとして使用すること
- WASMへの正式移行
- RL algorithm変更
- reward設計変更
- model architecture変更
- PPO / GAE

## 完了条件
親方針として、以下の実装群がsub-Issueを通じて成立することを目標とする。

- AI学習・評価専用のreusableなC++ Napoleon Core / Simulation Runtime
- RL一用途だけに依存しないAgent / Roster abstraction
- fixed roster / opponent pool / self-play
- TS game-coreとのrandomized differential/parity test
- deterministic seed semantics
- RuleBased + multiple ONNX policies
- policy別batched CUDA inference
- current×1 + RuleBased/v740 pool
- tensor-ready RL dataset生成
- AI evaluation / benchmarkへの再利用
- 2000-game benchmarkで現行TS runtimeと比較
- correctness regressionなし

## 関連
- #14 playing RL
- #104 rollout optimization
- #116 ONNX CUDA
- #118 tensor-ready dataset
- #120 ephemeral raw cache
- #121 dynamic batched CUDA rollout
- #125 batched CUDA behavior parity
